[Gatsby Docs Update] Menu overlay: scrolling fixes

joecritch opened the pull request:
**issues**

- when scrolling within the menu overlay, the background would capture the touch event instead, and you'd be scrolling the main content rather than the overlay's scrollbar. this would happy fairly randomly.
- a similar issue would happen if you scrolled to the top or bottom of the overlay (A.k.A. "overscroll")

_here's a demonstration of what was happening_

![ezgif-1-0be83948b1](https://user-images.githubusercontent.com/24449/30777375-08e4f4d8-a0b1-11e7-9482-98e3ce404fd9.gif)

**what's included**

- fixes for the both, by using `evt.preventDefault()` on the main content within the docs, and then taking care of the header and footer by hiding them. (The header actually animates away, whereas the footer can never be seen, so that isn't applicable.).
- "overscroll" has been taken care of with a fix inspired by this (http://blog.christoffer.online/2015-06-10-six-things-i-learnt-about-ios-rubberband-overflow-scrolling/). (more idiomatic to React though)

_screencast preview of the fixes:_

https://cl.ly/1f0T2G452239

cc @bvaughn @flarnie

joecritch commented on www/src/components/MarkdownPage/MarkdownPage.js:
```diff
@@ -21,102 +21,146 @@ import StickyResponsiveSidebar from 'components/StickyResponsiveSidebar';
+        isMenuOverlayOpen: willOpen,
+      },
+      () => {
+        this.context.onMenuOverlayToggle(this.state.isMenuOverlayOpen);
```
Because Gatsby's layout can't re-render its children, this instead uses context to notify the layout to "sync" the header and footer visibility, based on whether than menu is open.

flarnie replied:
> Because Gatsby's layout can't re-render its children

I believe you, and I'm also curious about this. Won't dig into it much for now but would be curious if you have time to comment more about that.

This seems a bit complicated to me. Going to read through it again and make sure I understand what is happening with this context call.

flarnie replied:
I read through this again, and I'm still not sure why the `onMenuOverlayToggle` is not passed down via a prop instead of context? Then we could call `this.props.onMenuOverlayToggle` here. Context is fine too, but more obscure than just using props I think.

joecritch replied:
I'd be keen on using props instead of context, for sure. I'm just not sure how this would be implemented? Would we pass arguments to `children()`?

Also, here's some notes about this setState callback architecture:

At the minute, Gatsby has its own sCU implementation in its route-level components. This means that, if we set some layout-level state, the route-level components (i.e. `<Docs>`) wouldn't re-render. So we have to work around this by keeping the state inside of `<Docs>`, and communicating back to the layout with context. (Which in turn has it's own "synced" state). Not ideal, but the best way I can think of atml

flarnie replied:
I see - didn't realize this was a limitation of Gatsby, that we can't pass custom props to the children of the template. Makes sense!

flarnie commented:
Hi Joe - Brian has said similar things on the other two PRs, but wanted to flag that I'm still seeing the issue where tapping the navbar on mobile doesn't open the navigation menu. I saw this in both iOS and my Samsung android phone. I was using the netlify link.

Brian and I are a bit stumped about what is causing this issue.

flarnie commented on www/src/components/MarkdownPage/MarkdownPage.js:
```diff
@@ -21,102 +21,146 @@ import StickyResponsiveSidebar from 'components/StickyResponsiveSidebar';
+      // If the screen has been resized >= 'small'
+      this.toggleMenuOverlay(false);
+    }
+  }
```
super-nit: I usually put newlines between method definitions. Not sure we are doing that in the docs code though.

joecritch replied:
Good catch, I'll do that now

flarnie commented on www/src/components/MarkdownPage/MarkdownPage.js:
```diff
@@ -21,102 +21,146 @@ import StickyResponsiveSidebar from 'components/StickyResponsiveSidebar';
+  handleTouchMove(evt) {
+    // Prevent anything else from being touched when the menu overlay is open, to avoid touch scrolling issues
+    if (this.state.isMenuOverlayOpen) {
+      evt.preventDefault();
```
This seems good, but it also makes me nervous;
- how can we ensure that this will not catch `touchMove` events from within the menu overlay?
- what if we somehow get into a state where `isMenuOverlayOption` is true but the user can't see the menu overlay? Can we add an invariant somehow to prevent that?

I know these are not great suggestions - will try to come up with something more specific on my second pass at this.

joecritch replied:
I feel ya! But it's unfortunately the only fix I've found for the drastic Safari scrolling bugs.

- It doesn't catch touch events from the menu, because of the corresponding `evt.stopPropagation()` I've added.
- I don't see how the user would see the menu overlay, as that's controlled by the open prop, which is derived from the state itself. Can you elaborate? :)

flarnie commented on www/src/components/StickyResponsiveSidebar/StickyResponsiveSidebar.js:
```diff
@@ -13,27 +13,43 @@
+
+  handleMenuTouchMove(evt) {
+    // Prevent evt.preventDefault in <MarkdownPage />
+    evt.stopPropagation();
```
Again - this makes me a bit nervous, so I might take some time to think about possible problems we might run into by catching this event here.

flarnie commented on www/src/components/StickyResponsiveSidebar/StickyResponsiveSidebar.js:
```diff
@@ -13,27 +13,43 @@
+  }
+
+  handleMenuTouchStart(evt) {
+    // Prevent "overscrolling"
```
This does make sense - as a follow-up it would be nice to isolate in a wrapper component if possible, but no big deal for now.

joecritch commented on www/src/components/LayoutFooter/Footer.js:
```diff
@@ -20,13 +20,14 @@ import {colors, media} from 'theme';
       color: colors.white,
       paddingTop: 10,
       paddingBottom: 50,
+      display: isVisible === false ? 'none' : null,
```
Yeah, I feel like this is most idiomatic with Glamor (as opposed to using global class names, etc.)

joecritch commented on www/src/components/MarkdownPage/MarkdownPage.js:
```diff
@@ -21,102 +21,146 @@ import StickyResponsiveSidebar from 'components/StickyResponsiveSidebar';
+  componentDidUpdate(prevProps, prevState) {
+    if (this.state.isMenuOverlayOpen && !prevState.isMenuOverlayOpen) {
+      // Menu has been opened
+      window.addEventListener('resize', this.handleWindowResize);
```
Good spot. Even though it'd be hard for the user to exit the `<Docs />` page without closing the menu (I think?), this is definitely good to have anyway.

joecritch commented on www/src/components/MarkdownPage/MarkdownPage.js:
```diff
@@ -21,102 +21,146 @@ import StickyResponsiveSidebar from 'components/StickyResponsiveSidebar';
+  }
+  handleWindowResize() {
+    const mediaQuery = media.greaterThan('small').replace('@media ', '');
+    if (window.matchMedia(mediaQuery).matches) {
```
To be honest, the docs haven't been tested in IE9. I was assuming IE11+ 👍 Do let me know if that's an issue.

bvaughn replied:
React itself aims to support IE9+ (see [here](https://facebook.github.io/react/docs/react-dom.html#browser-support)) but I think it's okay if the docs site only targets 11+.

bvaughn commented:
I'll defer to Flarnie's judgement on this PR b'c I've been too busy to dig in 😄

bvaughn commented on www/src/components/MarkdownPage/MarkdownPage.js:
```diff
@@ -21,101 +21,154 @@ import StickyResponsiveSidebar from 'components/StickyResponsiveSidebar';
+    if (this.state.isMenuOverlayOpen && !prevState.isMenuOverlayOpen) {
+      // Menu has been opened
+      window.addEventListener('resize', this.handleWindowResize);
+    }
```
Should probably clean-up on close too:
```js
if (this.state.isMenuOverlayOpen && !prevState.isMenuOverlayOpen) {
  // add handler
} else if (!this.state.isMenuOverlayOpen && prevState.isMenuOverlayOpen) {
  // remove handler
}
```
B'c not all closes will reload the page (eg Tutorial section). Or we could just get rid of the previous check entirely:
```js
if (this.state.isMenuOverlayOpen) {
  // add handler
} else {
  // remove handler
}
```

bvaughn commented:
I noticed a bug on the Netlify preview site ☹️ : https://deploy-preview-10792--reactjs.netlify.com/docs/hello-world.html

After clicking a link, the header+footer stay permanently hidden (unless you reload the page). Here's a repro:

![mobile-nav-bug](https://user-images.githubusercontent.com/29597/30835694-6229770e-a20e-11e7-8e1f-9c4866edae1a.gif)

flarnie commented:
I'm going to check and see if the behavior is different in netlify vs. when I build and serve locally.

flarnie commented:
Yea, sorry to confirm but this was reproducible locally with Chrome as well - the same stuff Brian saw.

joecritch commented:
@bvaughn @flarnie Thanks for the reviews, both :) Yeah, these were legit component state issues. Should all be fixed now. Also added a slight improvement to animation perf, and ensured Installation and Error Decoder follow suit, by abstracted everything menu overlay related into `<MenuOverlayContainer>`.

bvaughn commented on www/src/components/MenuOverlayContainer/MenuOverlayContainer.js:
```diff
@@ -0,0 +1,79 @@
+    );
+  }
+
+  handleTouchMove(evt) {
```
This method needs to be bound to the instance. Else it errors during runtime:

1. Open the mobile nav
1. Click the "X" to close it
1. Touch and scroll the page to drag
1. Console error:
```
Uncaught TypeError: Cannot read property 'state' of undefined
  at t.handleTouchMove
```

Fix should be either to call `.bind` in the constructor (like you do for the other class methods) or use this syntax:
```js
handleTouchMove = (evt) => {
  // ...
};
```

bvaughn commented:
I see a warning in the console on this branch that I don't see on `gatsby` branch:
```
Warning: Unsupported style property @media (min-width: 600px). Did you mean @media (minWidth: 600px)?

Check the render method of `StickyResponsiveSidebar`.
```

joecritch commented:
@bvaughn Appreciate the input. Everything should be fixed up now. Also rebased recent gatsby branch changes, removed the margin-top transition jank, and a "q" typo.

bvaughn commented:
> _here's a demonstration of what was happening_

Btw, I'm curious if you're still able to reproduce the behavior shown in the PR description video on the live site? I'm not.

joecritch commented:
@bvaughn Yeah, if my touch starts on either the header, or the below the nav items. Doesn't seem _too_ bad, if we perhaps could figure out how to put the footer under the nav.

bvaughn commented:
Gotcha! I'm not able to reproduce the footer overlap issue. And the header remaining visible doesn't seem that bad to me (since we have empty white space up top in this branch too). Hm.

joecritch commented:
Closing this after a discussion away from GitHub. Has a lot of moving parts for something that is quite an edge-case in the user flow
